### PR TITLE
ScaLAPACKMatrix: add functions to perform addition, multiplication and row/column scaling

### PR DESCRIPTION
**Pull request by BenBrands**

### CODE REVIEW

File: include/deal.II/lac/scalapack.h [hunk 1]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
                const std::pair<unsigned int,unsigned int> &offset_B,
                const std::pair<unsigned int,unsigned int> &submatrix_size) const;
 
+  /**
+   * Transposing assignment: <i>A = B<sup>T</sup></i>
+   *
+   * The matrices <tt>A</tt> and <tt>B</tt> must have the same process grid
+   *
+   * Following alignment conditions have to be fulfilled: MB_A = NB_B and NB_A = MB_B
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: in class description we have

> ...into $MB$ by $NB$ blocks...

not sure how it renders, but probably you should do $MB_A$ and alike here and below in this PR to keep it consistent.

File: include/deal.II/lac/scalapack.h [hunk 2]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
                const std::pair<unsigned int,unsigned int> &offset_B,
                const std::pair<unsigned int,unsigned int> &submatrix_size) const;
 
+  /**
+   * Transposing assignment: <i>A = B<sup>T</sup></i>
+   *
+   * The matrices <tt>A</tt> and <tt>B</tt> must have the same process grid
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: full stop

File: include/deal.II/lac/scalapack.h (hunk 1 above)
Comment (davydden) [reply include/deal.II/lac/scalapack.h]: full stop

File: include/deal.II/lac/scalapack.h [hunk 3]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+  /**
+   * Matrix-addition:
+   *
+   * <i>A = a A + b op(B)</i>
+   *
+   * if(transpose_B)  <i>op(B) = B<sup>T</sup></i>
+   *
+   * else  <i>op(B) = B</i>
+   *
+   * The matrices <tt>A</tt> and <tt>B</tt> must have the same process grid
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: full stop

File: include/deal.II/lac/scalapack.h [hunk 4]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   *
+   * Following alignment conditions have to be fulfilled:
+   *
+   * | transpose_B |          Block Sizes         |
+   * | :---------: | :--------------------------: |
+   * |   false     | MB_A = MB_B <br> NB_A = NB_B |
+   * |   true      | MB_A = NB_B <br> NB_A = MB_B |
+   */
+  void add(const ScaLAPACKMatrix<NumberType> &B,
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: now i see what you wanted to implement, elsewhere (for vector classes) we call this:

```
sadd (const Number s, const Number a, const VectorSpaceVector< Number > & V)
```

since other matrix classes don't support this operation, i would rather stick with

```
sadd (const .. s, const ... a,const ... B, const bool transpose_B = false)
```

@dealii/dealii any opinions on naming of this matrix operation?

File: include/deal.II/lac/scalapack.h [hunk 5]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * | false       |   true      | MB_A = MB_C <br> NB_A = NB_B <br> MB_B = NB_C |      <i>C = a A * B<sup>T</sup> + b C</i>        |
+   * | true        |   false     | MB_A = MB_B <br> NB_A = MB_C <br> NB_B = NB_C |        <i>C = a A<sup>T</sup> * B + b C</i>      |
+   * | true        |   true      | MB_A = NB_B <br> NB_A = MB_C <br> MB_B = NB_C | <i>C = a A<sup>T</sup> * B<sup>T</sup> + b C</i> |
+   *
+   * It is assumed that <tt>A</tt> and <tt>B</tt> have compatible sizes and that
+   * <tt>C</tt> already has the right size.
+   *
+   * The matrices <tt>A</tt>, <tt>B</tt> and <tt>C</tt> must have the same process grid.
+   */
+  void mult(ScaLAPACKMatrix<NumberType> &C,
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: @dealii/dealii this is kind of `smult` (scale matrix and add multiplication). For consistency with vectors, i would probably move two numbers in front i.e. `(const NumberType s, const NumberType a, const Matrix....)`.
But then it would be inconsistent with `FullMatrix`

```
void mmult (FullMatrix< number2 > &C, const FullMatrix< number2 > &B, const bool adding=false) const;
```

hm...

File: include/deal.II/lac/scalapack.h [hunk 6]
```diff
@@ -290,29 +447,25 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
   * It is assumed that $A$ has full rank: $rank(A) = \min(M,N)$.
   *
   * The following options are supported:
-  * - 1. If <code>transpose==false</code> and $M \geq N$: least squares solution of overdetermined system
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: since this is a separate stand-alone commit, can you please cherry-pick it into a stand-alone PR/branch?

File: include/deal.II/lac/scalapack.h [hunk 7]
```diff
@@ -389,6 +542,16 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
    */
   NumberType &local_el(const unsigned int loc_row, const unsigned int loc_column);
 
+  /**
+   * scaling the columns of the distributed matrix by scalars in array @p factors
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: Scale.... factors (full stop).

what is the size of `factors`?

File: include/deal.II/lac/scalapack.h [hunk 8]
```diff
@@ -389,6 +542,16 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   */
+  void scale_columns(std::vector<NumberType> &factors);
+
+  /**
+   * scaling the rows of the distributed matrix by scalars in array @p factors
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: Scale ... factors (full stop)

File: include/deal.II/lac/scalapack.templates.h [hunk 9]
```diff
@@ -1488,6 +1549,109 @@ inline void pgels(const char *trans,
   psgels_(trans,m,n,nrhs,A,ia,ja,desca,B,ib,jb,descb,work,lwork,info);
 }
 
+
+template <typename number>
+inline void pgeadd(const char */*transa*/,
```
Comment (davydden) [inline include/deal.II/lac/scalapack.templates.h]: please add space between `*` and comment block. same below.

File: source/lac/scalapack.cc [hunk 10]
```diff
@@ -439,6 +439,193 @@ ScaLAPACKMatrix<NumberType>::copy_to (ScaLAPACKMatrix<NumberType> &dest) const
...
+    {
+      Assert(this->n_rows==B.n_columns,ExcDimensionMismatch(this->n_rows,B.n_columns));
+      Assert(this->n_columns==C.n_rows,ExcDimensionMismatch(this->n_columns,C.n_rows));
+      Assert(B.n_rows==C.n_columns,ExcDimensionMismatch(B.n_rows,C.n_columns));
+
+      Assert(this->column_block_size==C.row_block_size,ExcDimensionMismatch(this->row_block_size,C.row_block_size));
+      Assert(this->row_block_size==B.column_block_size,ExcDimensionMismatch(this->column_block_size,B.row_block_size));
+      Assert(B.row_block_size==C.column_block_size,ExcDimensionMismatch(B.column_block_size,C.column_block_size));
+    }
+  Threads::Mutex::ScopedLock lock (mutex);
```
Comment (davydden) [inline source/lac/scalapack.cc]: we need mutex only when `mutable` member variables like `work` array are being used. I think you don't need it here.

File: source/lac/scalapack.cc [hunk 11]
```diff
@@ -1320,6 +1507,46 @@ void ScaLAPACKMatrix<NumberType>::load_parallel(const char *filename)
...
 
+template <typename NumberType>
+void ScaLAPACKMatrix<NumberType>::scale_columns(std::vector<NumberType> &factors)
+{
+  Assert(n_columns==(int)factors.size(),ExcDimensionMismatch(n_columns,factors.size()));
+
+  if (this->grid->mpi_process_is_active)
+    {
+      for (int i=0; i<n_local_rows; ++i)
+        {
```
Comment (davydden) [inline source/lac/scalapack.cc]: don't need these brackets

File: source/lac/scalapack.cc [hunk 12]
```diff
@@ -1320,6 +1507,46 @@ void ScaLAPACKMatrix<NumberType>::load_parallel(const char *filename)
...
+          for (int j=0; j<n_local_columns; ++j)
```
Comment (davydden) [inline source/lac/scalapack.cc]: you should probably loop over columns first, take value and multiply all row elements.

File: source/lac/scalapack.cc [hunk 13]
```diff
@@ -1320,6 +1507,46 @@ void ScaLAPACKMatrix<NumberType>::load_parallel(const char *filename)
...
+  Assert(n_rows==(int)factors.size(),ExcDimensionMismatch(n_rows,factors.size()));
+
+  if (this->grid->mpi_process_is_active)
+    {
+      for (int i=0; i<n_local_rows; ++i)
+        {
+          const int glob_i = global_row(i);
+          for (int j=0; j<n_local_columns; ++j)
+            {
+              local_el(i,j) *= factors[glob_i];
```
Comment (davydden) [inline source/lac/scalapack.cc]: just do `const NumberType & s = factors[global_row(i)];`

File: source/lac/scalapack.cc [hunk 14]
```diff
@@ -1320,6 +1507,46 @@ void ScaLAPACKMatrix<NumberType>::load_parallel(const char *filename)
...
+{
+  Assert(n_rows==(int)factors.size(),ExcDimensionMismatch(n_rows,factors.size()));
+
+  if (this->grid->mpi_process_is_active)
+    {
+      for (int i=0; i<n_local_rows; ++i)
+        {
+          const int glob_i = global_row(i);
+          for (int j=0; j<n_local_columns; ++j)
+            {
```
Comment (davydden) [inline source/lac/scalapack.cc]: and remove these brackets

File: include/deal.II/lac/scalapack.h (hunk 1 above)
Comment (BenBrands) [reply include/deal.II/lac/scalapack.h]: I will change it at the one point in the class description, if that is fine with you.

Comment (davydden) [reply include/deal.II/lac/scalapack.h]: i would actually prefer the opposite, i think it will render better as latex formula. Otherwise you have it as a text `MB_A`

Comment (davydden) [reply include/deal.II/lac/scalapack.h]: p.s. that's also how we do it all other places inside ScaLAPACK, including the most recent PR with some docu changes https://github.com/dealii/dealii/pull/5875/

Comment (BenBrands) [reply include/deal.II/lac/scalapack.h]: Right ...

**Comment by davydden** [comment]
@dealii/dealii can someone please have a look at the documentation here and the chosen names/arguments for member functions?

File: include/deal.II/lac/scalapack.h [hunk 15]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
                const std::pair<unsigned int,unsigned int> &offset_B,
                const std::pair<unsigned int,unsigned int> &submatrix_size) const;
 
+  /**
+   * Transposing assignment: <i>A = B<sup>T</sup></i>
+   *
+   * The matrices <tt>A</tt> and <tt>B</tt> must have the same process grid.
+   *
+   * Following alignment conditions have to be fulfilled: $MB_A$ = $NB_B$ and $NB_A$ = $MB_B$
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.h]: Say "The following ...". I had to think about whether "following" is a verb or adjective. Same in a few places below.

Also end the sentence with a period.

File: include/deal.II/lac/scalapack.h [hunk 16]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   */
+  void copy_transposed(const ScaLAPACKMatrix<NumberType> &B);
+
+  /**
+   * Matrix-addition:
+   *
+   * <i>A = a A + b op(B)</i>
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.h]: remove the empty line so that this reads back to back

think about whether you could do the same in the following lines

File: include/deal.II/lac/scalapack.h [hunk 17]
```diff
@@ -389,6 +546,24 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
   NumberType &local_el(const unsigned int loc_row, const unsigned int loc_column);
 
+  /**
+   * Scale the columns of the distributed matrix by scalars in array @p factors.
+   *
+   * The array @p factors must have as many entries as the matrix columns.
+   *
+   * Copies of @p factors have to available on all processes of the underlying MPI communicator
+   */
+  void scale_columns(std::vector<NumberType> &factors);
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.h]: make the argument `const`

File: include/deal.II/lac/scalapack.h [hunk 18]
```diff
@@ -389,6 +546,24 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+
+  /**
+   * Scale the rows of the distributed matrix by scalars in array @p factors.
+   *
+   * The array @p factors must have as many entries as the matrix rows.
+   *
+   * Copies of @p factors have to available on all processes of the underlying MPI communicator
+   */
+  void scale_rows(std::vector<NumberType> &factors);
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.h]: same here

File: source/lac/scalapack.cc [hunk 19]
```diff
@@ -439,6 +439,192 @@ ScaLAPACKMatrix<NumberType>::copy_to (ScaLAPACKMatrix<NumberType> &dest) const
...
+                                       const bool transpose_B) const
+{
+  Assert(this->grid==B.grid,ExcMessage("The matrices A and B need to have the same process grid"));
+  Assert(C.grid==B.grid,ExcMessage("The matrices B and C need to have the same process grid"));
+
+  // see for further info:
+  // https://www.ibm.com/support/knowledgecenter/SSNR5K_4.2.0/com.ibm.cluster.pessl.v4r2.pssl100.doc/am6gr_lgemm.htm
+  if (!transpose_A && !transpose_B)
+    {
+      Assert(this->n_columns==B.n_rows,ExcDimensionMismatch(this->n_columns,B.n_rows));
```
Comment (bangerth) [inline source/lac/scalapack.cc]: thanks for adding all of these asserts!

**Comment by jppelteret** [comment]
@BenBrands I'll take a look through the documentation tomorrow!

File: include/deal.II/lac/scalapack.h (hunk 15 above)
Comment (jppelteret) [reply include/deal.II/lac/scalapack.h]: You beat me to it! I was also going to make the suggestion that the style of all of the documentation be made more consistent with that of the rest of the class. The conditions can also be put in a `@note`.

**Comment by jppelteret** [comment]
Ok, so @bangerth happened post an initial few comment just commented, but I'll stand by my initial commitment and will give it a proper look tomorrow.

File: include/deal.II/lac/scalapack.h (hunk 15 above)
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: I think that for all of the functions that you're documented, the style could be made a bit more similar to that of the existing document. So here, for example, you could consider writing it as follows:
```
Copy the transpose of the contents of the distributed matrix @p B into this one.

This transpose assignment operation modifies the calling matrix @p A such 
that $A = B^{T}$.

@note The matrices @p A and @p B must have the same process grid.
The following alignment conditions also have to be fulfilled: 
$MB_A = NB_B$ and $NB_A = MB_B$.
```

File: include/deal.II/lac/scalapack.h [hunk 20]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
                const std::pair<unsigned int,unsigned int> &offset_B,
                const std::pair<unsigned int,unsigned int> &submatrix_size) const;
 
+  /**
+   * Transposing assignment: <i>A = B<sup>T</sup></i>
+   *
+   * The matrices <tt>A</tt> and <tt>B</tt> must have the same process grid.
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: Why have you used HTML as opposed to LaTeX formatting (both here and below)?

File: include/deal.II/lac/scalapack.h [hunk 21]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+  void copy_transposed(const ScaLAPACKMatrix<NumberType> &B);
+
+  /**
+   * Matrix-addition:
+   *
+   * <i>A = a A + b op(B)</i>
+   *
+   * if(transpose_B)  <i>op(B) = B<sup>T</sup></i>
+   *
+   * else  <i>op(B) = B</i>
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: I'd prefer for this to be expressed in words rather than as pseudo-code. i.e.

```
If the @p transpose_B is set then the transpose of @p B is used.
That is to say that the operation $op(B)$ corresponds to $B^{T}$. 
If this flag is set to false, then $op(B)$ applies @p B with its default ordering.
```

This also applies below.

File: include/deal.II/lac/scalapack.h (hunk 15 above)
Comment (jppelteret) [reply include/deal.II/lac/scalapack.h]: Note also that $MB_A$ = $NB_B$ should be $MB_A = NB_B$ so that the equals is also rendered using LaTeX. Same applies below.

File: include/deal.II/lac/scalapack.h (hunk 5 above)
Comment (jppelteret) [reply include/deal.II/lac/scalapack.h]: We used to have similar functions in some of the vector classes (e.g [here](https://www.dealii.org/8.5.0/doxygen/deal.II/classTrilinosWrappers_1_1VectorBase.html#ab0d14dd0a857a51f18562e84a989710c)), so I would say that the factors and matrices should be ordered according to the operation along with the input matrices, even if that means that we lose the default values. Does a default value of 0 for `b` make sense anyway? This then just becomes an assignment operation.

File: include/deal.II/lac/scalapack.h [hunk 22]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+            const ScaLAPACKMatrix<NumberType> &B,
+            const NumberType a=1,
+            const NumberType b=0,
+            const bool transpose_A=false,
+            const bool transpose_B=false) const;
+
+  /**
+   * Matrix-matrix-multiplication.
+   *
+   * The optional parameter <tt>adding</tt> determines, whether the result is
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: no comma

File: include/deal.II/lac/scalapack.h [hunk 23]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * stored in <tt>C</tt> or added to <tt>C</tt>.
+   *
+   * if (adding) <i>C += A*B</i>
+   *
+   * else <i>C = A*B</i>
+   *
+   * Assumes that <tt>A</tt> and <tt>B</tt> have compatible sizes and that
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: It is assumed that...

File: include/deal.II/lac/scalapack.h [hunk 24]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * Following alignment conditions have to be fulfilled: $MB_A$ = $MB_C$, $NB_A$ = $MB_B$ and $NB_B$ = $NB_C$
+   */
+  void mmult(ScaLAPACKMatrix<NumberType> &C,
+             const ScaLAPACKMatrix<NumberType> &B,
+             const bool adding=false) const;
+
+  /**
+   * Matrix-matrix-multiplication using transpose of <tt>this</tt>.
+   *
+   * The optional parameter <tt>adding</tt> determines, whether the result is
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: no comma

File: include/deal.II/lac/scalapack.h [hunk 25]
```diff
@@ -189,6 +189,163 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * stored in <tt>C</tt> or added to <tt>C</tt>.
+   *
+   * if (adding) <i>C += A<sup>T</sup>*B</i>
+   *
+   * else <i>C = A<sup>T</sup>*B</i>
+   *
+   * Assumes that <tt>A</tt> and <tt>B</tt> have compatible sizes and that
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: It is assumed that...

File: include/deal.II/lac/scalapack.h (hunk 20 above)
Comment (jppelteret) [reply include/deal.II/lac/scalapack.h]: You should use `@p` to format the parameters of a function, i.e. `@p A` and `@p B`, when inline with other text.

Same applies below.

File: include/deal.II/lac/scalapack.h [hunk 26]
```diff
@@ -389,6 +546,24 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
    */
   NumberType &local_el(const unsigned int loc_row, const unsigned int loc_column);
 
+  /**
+   * Scale the columns of the distributed matrix by scalars in array @p factors.
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: by the scalars provided in the array

File: include/deal.II/lac/scalapack.h [hunk 27]
```diff
@@ -389,6 +546,24 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   *
+   * The array @p factors must have as many entries as the matrix columns.
+   *
+   * Copies of @p factors have to available on all processes of the underlying MPI communicator
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: have to be; fullstop at end of sentence.

File: include/deal.II/lac/scalapack.h [hunk 28]
```diff
@@ -389,6 +546,24 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   */
+  void scale_columns(std::vector<NumberType> &factors);
+
+  /**
+   * Scale the rows of the distributed matrix by scalars in array @p factors.
+   *
+   * The array @p factors must have as many entries as the matrix rows.
+   *
+   * Copies of @p factors have to available on all processes of the underlying MPI communicator
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: have to be; fullstop at end of sentence.

File: include/deal.II/lac/scalapack.h [hunk 29]
```diff
@@ -389,6 +546,24 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * Scale the columns of the distributed matrix by scalars in array @p factors.
+   *
+   * The array @p factors must have as many entries as the matrix columns.
+   *
+   * Copies of @p factors have to available on all processes of the underlying MPI communicator
+   */
+  void scale_columns(std::vector<NumberType> &factors);
+
+  /**
+   * Scale the rows of the distributed matrix by scalars in array @p factors.
```
Comment (jppelteret) [inline include/deal.II/lac/scalapack.h]: by the scalars provided in the array

**Comment by jppelteret** [comment]
Thanks for taking the time to add this functionality @BenBrands! Since @davydden checked the implementation already, lets clean up the documentation a little bit and then I think that this is good to go!

File: include/deal.II/lac/scalapack.h (hunk 18 above)
Comment (davydden) [reply include/deal.II/lac/scalapack.h]: btw, we should probably use `Vector<NumberType>` instead here. It won't make much of a difference at the end of the day, but is probably more consistent representation of a diagonal matrix used here as pre/post multiplication.

File: include/deal.II/lac/scalapack.h (hunk 20 above)
Comment (davydden) [reply include/deal.II/lac/scalapack.h]: I am also in favour of LaTeX, but we have a mixture of the two, apparently https://github.com/dealii/dealii/issues/5897

File: include/deal.II/lac/scalapack.h (hunk 21 above)
Comment (BenBrands) [reply include/deal.II/lac/scalapack.h]: I would favor the pseudo-code as  it is short and precise.
But I can adapt it if wanted.

File: include/deal.II/lac/scalapack.h (hunk 15 above)
Comment (BenBrands) [reply include/deal.II/lac/scalapack.h]: I discussed with @davydden  the documentation of `ScaLAPACKMatrix` and we agree that it has to be reworked to make it uniform. I will do that in a different PR based on the style used for the doc in this PR.

File: include/deal.II/lac/scalapack.h (hunk 18 above)
Comment (davydden) [reply include/deal.II/lac/scalapack.h]: we may seek just a few eigenvalues or all eigenvalues of a matrix. In other places (like Arpack for sparse large matrices) we also use `std::vector<>` which is problably ok.

Here, however, this vector has to have as many elements as matrix rows/columns. So in case of a square matrix, this is simply a diagonal matrix represented by a vector. With other matrix classes AFAIK we never use `std::vector<>` to represent a diagonal, i.e. https://www.dealii.org/developer/doxygen/deal.II/classSparseMatrix.html#a07ba942c5b7466b3895fed3104fbcc26

**Comment by BenBrands** [comment]
Thank you all for your comments.
The documentation for this class needs an update, I would like to do so in a different PR (based on the doc in this PR and #5898).

File: include/deal.II/lac/scalapack.h (hunk 21 above)
Comment (jppelteret) [reply include/deal.II/lac/scalapack.h]: I suppose that the pseudo-code is consistent with [`LapackFullMatrix`](https://www.dealii.org/developer/doxygen/deal.II/classLAPACKFullMatrix.html#a2fb96068b464f9eaddccd74696bdaf42) and elements of the [`Vector` class](https://www.dealii.org/developer/doxygen/deal.II/classLinearAlgebra_1_1Vector.html#a5c4190172e7ebc876dbdcde8617813f3). Although I'm not a huge fan of it, I'm not going to push the issue since there's precedent for it.

**Comment by BenBrands** [comment]
The functions `scale_columns()` and  `scale_rows()` take `const ArrayView<const NumberType>` as input!

**Comment by jppelteret** [comment]
@BenBrands Thanks again for the contribution. I've created an issue for the proposed documentation enhancements.

**Comment by jppelteret** [comment]
@davydden @BenBrands Lets discuss the changes made to `scale_columns` and `scale_rows` before we merge this. This did not work out quite as I thought it would. I guess that we would need to more closely follow the idiom used [here](https://github.com/dealii/dealii/blob/master/include/deal.II/fe/fe_values.h#L355) and [here](https://github.com/dealii/dealii/blob/master/source/fe/fe_values.cc#L1365) to achieve it.

File: include/deal.II/lac/scalapack.h [hunk 30]
```diff
@@ -189,6 +189,153 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * The following alignment conditions have to be fulfilled: $MB_A=NB_B$ and $NB_A=MB_B$.
+   */
+  void copy_transposed(const ScaLAPACKMatrix<NumberType> &B);
+
+  /**
+   * The operations based on the input parameter @p transpose_B and the alignment conditions are summarized in the following table:
+   *
+   * | transpose_B |          Block Sizes         |                    Operation                  |
+   * | :---------: | :--------------------------: | :-------------------------------------------: |
+   * |   false     | $MB_A=MB_B$ <br> $NB_A=NB_B$ |  $\mathbf{A} = a \mathbf{A} + b \mathbf{B}$   |
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: in another PR https://github.com/dealii/dealii/pull/5899#discussion_r168256838 @kronbichler raised the point whether we shall use bold symbols (`\mathbf`) for matrices and vectors. 

Personally, I don't mind sticking with plain LaTeX `A`, `B`, `c` for linear algebra, probably with `\cdots` to denote multiplication. Downside is that we won't distinguish between scalars and vectors in this notation.

@dealii/dealii thoughts?

File: include/deal.II/base/array_view.h [hunk 31]
```diff
@@ -524,6 +524,46 @@ make_array_view (ElementType *const begin, ElementType *const end)
...
+ * It only exists for compatibility purposes.
+ *
+ * @param[in] array_view The ArrayView that we wish to make a copy of.
+ *
+ * @relates ArrayView
+ */
+template <typename Number>
+inline
+ArrayView<const Number>
+make_array_view (const ArrayView<Number> &array_view)
```
Comment (jppelteret) [inline include/deal.II/base/array_view.h]: @masterleinad I think that you added some of the other `make_array_view` functions previously. Are you happy with these two additions (they help instantiation of some functions)?

Comment (masterleinad) [reply include/deal.II/base/array_view.h]: Sure. In the end, I prefer to have external `make_array_functions` instead of modifying the class `ArrayView` itself. If this helps with compatibility, I am totally fine with it.

Comment (jppelteret) [reply include/deal.II/base/array_view.h]: Great, thanks for the confirmation!

**Comment by jppelteret** [comment]
Ok, so I think that this is ready to merge but since I've contributed directly to this PR it would be more appropriate for someone else to make the final call.

File: tests/scalapack/scalapack_13.cc [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+
+  const unsigned int proc_rows = std::floor(std::sqrt(n_mpi_processes));
+  const unsigned int proc_columns = std::floor(n_mpi_processes/proc_rows);
+  //create 2d process grid
+  std::shared_ptr<Utilities::MPI::ProcessGrid> grid = std::make_shared<Utilities::MPI::ProcessGrid>(mpi_communicator,proc_rows,proc_columns);
+  pcout << "2D process grid: " << grid->get_process_grid_rows() << "x" << grid->get_process_grid_columns() << std::endl << std::endl;
+
+  const std::vector<unsigned int> sizes = {{400,500}};
+
+  // test A = alpha A + beta B
```
Comment (davydden) [inline tests/scalapack/scalapack_13.cc]: could you please split this test into 4 (each does different operation) and the next one into two (rows/cols). If/when something goes south, it's much easier to debug a small tests as opposed to a huge one which covers all the cases .

File: source/lac/scalapack.cc [hunk 33]
```diff
@@ -1320,9 +1506,72 @@ void ScaLAPACKMatrix<NumberType>::load_parallel(const char *filename)
...
+}
+
+
+
+template <typename NumberType>
+template <class InputVector>
+void ScaLAPACKMatrix<NumberType>::scale_columns(const InputVector &factors)
+{
+  internal::scale_columns(*this, make_array_view(factors),
+                          this->grid->mpi_process_is_active);
```
Comment (davydden) [inline source/lac/scalapack.cc]: minor nitpick, i think it's easier to read as
```
  if (this->grid->mpi_process_is_active)
    internal::scale_columns(...);
```
that is, when you move that `if` outside of internal function which actually does the job.
